### ISSUE TITLE
TypeScript: `paths` in `compilerOptions` not handled properly

### ISSUE THREAD

**Issue by jackTheRipper** [avajs/ava#1973]
## Description

I am currently using Ava for running my test written in TS for my micro-services. 

In my TS configuration, I am using the property `path` in `compilerOptions` for resolving some modules.

When running my unit-tests, Ava complains that it cannot find the modules I defined in the `paths`.

When `transpiling` and running my application, the paths are correctly resolved.

Please find below my configuration for Ava and TypeScript.

### Test Source

`tsconfig.json`:

```json
{
  "compilerOptions": {
    "module": "commonjs",
    "noImplicitAny": true,
    "removeComments": true,
    "target": "es6",
    "allowJs": true,
    "alwaysStrict": true,
    "diagnostics": true,
    "moduleResolution": "node",
    "noUnusedLocals": true,
    "noUnusedParameters": true,
    "sourceMap": true,
    "outDir": "./build",
    "baseUrl": ".",
    "paths": {
      "@config/*": [ "configuration/*" ]
    }
  },
  "include": [
    "src/*.ts",
    "test/*.ts"
  ],
  "exclude": [
    "index.js",
    "node_modules",
    "typings"
  ]
}
```

Import example in a TS file:

```typescript
import * as config from '@config/common';
```

### Error Message & Stack Trace

Here is the error message

```
7:
8: import { CHECKOUT_API_URL } from '@config/common.js';
9:

Error: Cannot find module '@config/common.js'
```

### Config

Copy the relevant section from `package.json`:

```json
"ava": {
    "files": [
      "test/*.spec.ts"
    ],
    "sources": [
      "src/**"
    ],
    "concurrency": "4",
    "failFast": false,
    "failWithoutAssertions": false,
    "compileEnhancements": false,
    "extensions": [
      "ts"
    ],
    "require": [
      "ts-node/register"
    ]
  }
```

### Command-Line Arguments

Ava command used:

```
ava
```

## Environment

```
Node 10.13
Ava 1.0.0-rc.1
Npm 6.4.1
TypeScript 3.1.6
```

**Comment by novemberborn** [comment]
With the above setup, `ts-node/register` is compiling your TypeScript files. I suppose it should compile `import * as config from '@config/common'` to `import * as config from '../configuration/common'` or something. Hard to say why it isn't, but I don't think this is related to AVA itself. You should be able to reproduce this in a separate script, e.g. `node -r ts-node/register repro.js` where `repro.js` requires a TypeScript file that then has a `@config` dependency.

(I'm closing this issue for housekeeping purposes, but let's keep the conversation going.)

**Comment by havenchyk** [comment]
@novemberborn is right, the root cause of this problem is not in AVA itself, but in `ts-node` nature.

tl;dr TypeScript only "understands" aliases, but it doesn't convert them to what we would like to have (real path).

You certainly can run tests with aliases, but you will have to use another tool or script that will convert aliases to normal paths in your js code. I used [module-alias](https://www.npmjs.com/package/module-alias), but you can choose any of packages or there is a babel plugin for it - it's up to you.

If you're still interested in working tests, I believe it would be enough to install https://www.npmjs.com/package/tsconfig-paths and add `tsconfig-paths/register` to `require` section of ava config. 

You can check the link https://github.com/TypeStrong/ts-node#loading-tsconfigjson, ts-node recommends `tsconfig-paths` for such case.

@jackTheRipper you wrote that aliases are resolved correctly. What do you mean by this? Because if I have `import * as something from '~src/mypath` it's converted to `...require('~src/mypath')`.

**Comment by mesqueeb** [comment]
I have a similar issue, but not with typescript, but regular JavaScript.

I have certain webpack aliases that I also clarified in `jsconfig.js` file in the root of my project like so:

```js
{
  "compilerOptions": {
    // This must be specified if "paths" is set
    "baseUrl": ".",
    // Relative to "baseUrl"
    "paths": {
      "~helpers/*": ["src/helpers/*"],
      "~router/*": ["src/router/*"],
      "~store/*": ["src/store/*"],
      "~modules/*": ["src/store/modules/*"],
      "~components/*": ["src/components/*"],
      "~mixins/*": ["src/components/mixins/*"],
      "~config/*": ["src/config/*"],
    }
  }
}
```

This makes sure VSCode shows JSDoc style documentation even when using alias imports.
Is it possible to make it so AVA also checks for aliases in `jsconfig.js`, or should I set some special AVA config?

I searched for "alias" in issues and readme, but couldn't find much on this.

**Comment by novemberborn** [comment]
@mesqueeb you need to make sure that Node.js resolves those aliases. This PR contains a solution, but I think it's in the wrong place. Perhaps you could give it a try and help out with documenting it? https://github.com/avajs/ava/pull/2071

**Comment by mesqueeb** [comment]
Oh wow, I came here 1 year later with the same issue as I had 1 year ago. 😅 

These days I test directly with the TypeScript files of my source code.

For anyone with this issue, just do:
```
npm i -D tsconfig-paths
```
and make your AVA config like so:
```
  "ava": {
    "extensions": [
      "ts"
    ],
    "require": [
      "tsconfig-paths/register",
      "ts-node/register"
    ],
```
problem solved!
🎉

**Comment by novemberborn** [comment]
Or if you happy to compile before running tests, our [`@ava/typescript`](https://github.com/avajs/typescript/) package works pretty well.
